src/common/sv2_noise.h

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) 2023-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_COMMON_SV2_NOISE_H
+#define BITCOIN_COMMON_SV2_NOISE_H
+
+#include <compat/compat.h>
+#include <crypto/poly1305.h>
+#include <key.h>
+#include <pubkey.h>
+#include <random.h>
+#include <streams.h>
+#include <uint256.h>
+
+/** The Noise Protocol Framework
+ *  https://noiseprotocol.org/noise.html
+ *  Revision 38, 2018-07-11
+ *
+ *  Stratum v2 handshake and cipher specification:
+ *  https://github.com/stratum-mining/sv2-spec/blob/main/04-Protocol-Security.md
+ */
+
+/** Section 3: All Noise messages are less than or equal to 65535 bytes in length. */
+static constexpr size_t NOISE_MAX_CHUNK_SIZE = 65535;
+/** Sv2 spec 4.5.2 */
+
+/** Simple certificate for the static key signed by the authority key.
+ * See 4.5.2 and 4.5.3 of the Stratum v2 spec.
+ */
+class Sv2SignatureNoiseMessage
+{
+public:
+    /** Size of a Schnorr signature. */
+    static constexpr size_t SCHNORR_SIGNATURE_SIZE = 64;
+    /** Size of serialized message, which does not include the static key.  */
+    static constexpr size_t SIZE = 2 + 4 + 4 + SCHNORR_SIGNATURE_SIZE;
+
+private:
+    uint16_t m_version = 0;
+    uint32_t m_valid_from = 0;
+    uint32_t m_valid_to = 0;
+    std::array<unsigned char, SCHNORR_SIGNATURE_SIZE> m_sig;
+
+    /** Hash of version, valid from/to and the static key. */
+    uint256 GetHash();
+    void SignSchnorr(const CKey& authority_key, Span<unsigned char> sig);
+
+public:
+    Sv2SignatureNoiseMessage() = default;
+    Sv2SignatureNoiseMessage(uint16_t version, uint32_t valid_from, uint32_t valid_to, const XOnlyPubKey& static_key, const CKey& authority_key);
+
+    /* The certificate serializes pubkeys in x-only format, not EllSwift. */
+    XOnlyPubKey m_static_key = {};
+
+    [[nodiscard]] bool Validate(XOnlyPubKey authority_key);
+
+    template <typename Stream>
+    // The static_key is signed for, but not serialized.
+    void Serialize(Stream& s) const
+    {
+        s << m_version
+          << m_valid_from
+          << m_valid_to
+          << m_sig;
+    }
+    template <typename Stream>
+    void Unserialize(Stream& s)
+    {
+        s >> m_version >> m_valid_from >> m_valid_to >> m_sig;
```

Comment:
styling nit: Consider splitting into multiple lines for clarity and consistency with Serialize above.